CircuitPython_Pyloton/pyloton.py

```diff
@@ -0,0 +1,563 @@
+        radio = adafruit_ble.BLERadio()
+        a = SolicitServicesAdvertisement()
+        a.solicited_services.append(AppleMediaService)
+        radio.start_advertising(a)
+
+        while not radio.connected:
+            pass
+
+        self._status_update("AppleMediaService: Connected")
+
+        #known_notifications = set()
+
+        for connection in radio.connections:
+            if not connection.paired:
+                connection.pair()
+                self._status_update("AppleMediaService: Paired")
+
+            self.ams = connection[AppleMediaService]
+
+        return radio
+
+
+    def speed_cad_connect(self):
+        """
+        Connects to speed and cadence sensor
+        """
+        self._status_update("Speed and Cadence: Scanning...")
+        # Save advertisements, indexed by address
+        advs = {}
+        for adv in self.ble.start_scan(ProvideServicesAdvertisement, timeout=5):
+            if CyclingSpeedAndCadenceService in adv.services:
+                self._status_update("Speed and Cadence: Found an advertisement")
+                # Save advertisement. Overwrite duplicates from same address (device).
+                advs[adv.address] = adv
+
+        self.ble.stop_scan()
+        self._status_update("Speed and Cadence: Stopped scanning")
+        if not advs:
+            # Nothing found. Go back and keep looking.
+            return []
+
+        # Connect to all available CSC sensors.
+        self.cyc_connections = []
+        for adv in advs.values():
+            self.cyc_connections.append(self.ble.connect(adv))
+            self._status_update("Speed and Cadence: Connected {}".format(len(self.cyc_connections)))
+
+
+        self.cyc_services = []
+        for conn in self.cyc_connections:
+
+            self.cyc_services.append(conn[CyclingSpeedAndCadenceService])
+        self._status_update("Pyloton: Finishing up...")
+        return self.cyc_connections
+
+
+    def _speed_helper(self, values):
+        wheel_diff = values.last_wheel_event_time - self._previous_wheel
+        rev_diff = values.cumulative_wheel_revolutions - self._previous_revolutions
+
+        if wheel_diff:
+            # Rotations per minute is 60 times the amount of revolutions since
+            # the last update over the time since the last update
+            rpm = 60*(rev_diff/(wheel_diff/1024))
+            # We then mutiply it by the wheel's circumference and convert it to mph
+            speed = round((rpm * self.circumference) * (60/63360), 1)
+            if speed < 0:
+                speed = self._previous_speed
+            self._previous_speed = speed
+            self._previous_revolutions = values.cumulative_wheel_revolutions
+            self._i = 0
+        else:
+            self._i += 1
+            if self._i >= 3:
+                speed = 0
+        self._previous_wheel = values.last_wheel_event_time
+
+
+    def _cad_helper(self, values):
```

Comment:
Would it make sense to call this `_compute_cadence_and_speed` or similar?

Reply:
Yeah, I think so